torch/_dynamo/side_effects.py

```diff
@@ -375,6 +380,56 @@ def codegen_save_tempvars(self, cg: PyCodegen):
                 ]
             )
 
+    def register_hook(self, tensor, hook, handle):
+        self.tensor_hooks.append((tensor, hook, handle))
+
+    def codegen_hooks(self, cg):
+        for (
+            tensor,
+            hook,
+            handle,
+        ) in self.tensor_hooks:
+            # Note: [On tensor.register_hook]
+            #
+            # register_hook on a tensor, AKA backward hooks, have slightly nuanced differences in how they are implemented
+            # when it comes to hooks on objects with sources (inputs, params) vs objects without sources (intermediaries).
```

Comment:
Is it worth explaining why these two cases need to be differentiated?

My understanding (lmk if this sounds wrong) is: hooks on intermediates and outputs can be captured directly by AOTAutograd, since they get sandwiched inside of the backward graph when AOTAutograd calls `autograd.grad(graph_outputs, graph_inputs)`.

But hooks on intermediates aren't supposed to run until **after** we've computed gradients for the graph inputs, which means that they need to run outside of AOTAutograd's backward graph. So instead, we rely on compiled autograd to first trace through AOTAutograd's captured backward graph, and then trace through any hooks on the grad_inputs (which dynamo installs in the residuals, here).

Reply:
The answer is even simpler :)

If you have an input, (or a param or global, but I lump those under inputs for this sake) - you have a source. The source is an effect, not a cause. The cause is that we know how to reference these objects in the user's invoking frame! Aka, they are a named or accessible object in f_locals/f_globals.

That means dynamo can, with relative ease, at the end of computing the graph for the forward, literally invoke:

`some_known_tensor.register_hook(...)` 

In the generated bytecode!

If you were to do that with an intermediary, you *still could* (in fact, an early prototype of this did that) but now you run afoul of preserving memory characteristics, because if you invoked `some_intermediary_tensor.register_hook(...)` in the generated bytecode, you have now forced `some_intermediary_tensor` to *not* be cleaned up / set to None, and have forced it to live alongside the rest of the graph AND get written out into outputs.

Seeing as the intermediary tensor is (a) fully encompassed within the graph (2) already has a clear lifecycle story - violating these things for the sake of hooks, and writing a bunch of messy code to preserve these structures is a no go.